keps/sig-network/20200204-cni-verification-rearchitecture.md

```diff
@@ -0,0 +1,579 @@
+
+ ```
+	builder := &utils.NetworkPolicySpecBuilder{}
+	builder = builder.SetName("allow-x-via-pod-and-ns-selector").SetPodSelector(map[string]string{"pod": "a"})
+	builder.SetTypeIngress()
+	builder.AddIngress(nil, &p80, nil, nil, map[string]string{"pod":"b"}, map[string]string{"ns":"y"}, nil, nil)
+
+	reachability := utils.NewReachability(allPods, true)
+	reachability.ExpectAllIngress(Pod("x/a"), false)
+	reachability.Expect(Pod("y/b"), Pod("x/a"), true)
+	reachability.Expect(Pod("x/a"), Pod("x/a"), true)
+  ```
+ This represents a significant reduction in code complexity, with the equivalent tests using the existing `network_policy.go` implementation being 3 to 4 times as long, mostly due to boilerplate around verification and go structures.
+
+*Further improvements to the testing API*
+
+- Make the function calls in network policy builder *even* more DSL-like, for example, 
+```
+Pod(...).InNamespace(...).CanAccess(...)
+```
+- Infer `Egress,Ingress` rules rather than force them to be specified, based on builder inputs.  They're redundant to begin with (i.e. calico doesn't even require them)
+- Add `From` and `To` semantics to the struct API calls in reachability.
+ 
+###  Note on Acceptance and Backwards compatibility
+
+Thus far there are two obvious ways to ensure backwards compatibility. 
+- Each old test will be converted to a truth table, first, as part of this work, so that the parity between old and new tests is clear and obvious 
+- Alternatively, these can be a next generation of policy tests which live in parallel to existing network policy tests for a release cycle, while they are vetted
+
+## Next steps
+
+As of now, network policy tests are not run regularly against any CNI.  Although we should not endorse one CNI over another, we should regularly validate
+that the NetworkPolicy tests *can* pass on *some* provider.  As part of this proposal, we propose commiting an annotation to the existing `network_policy.go` code which states, in clear and simple terms, what environment the `network_policy.go` test suite was run in, the last time which it was committed and passed.  It's also acceptable to commit this as a Markdown file in the documentation.
+ 
+There may be other, better ways of doing this.  Running an upstream validation job of these tests as a weekly PROW job, for example, would be a good way to make sure that these tests don't regress in the future.  This comes at the cost of coupling a job to an external CNI provider, so its not being explicitly suggested.
+
+## Other Improvement Ideas
```

Comment:
It would be good to validate that policy enforcement does not change for same-node and different-node cases. In other words, if you're testing connectivity between pods A and B, ensure that the results are the same regardless of whether or not A and B are on the same node.

Reply:
@sedefsavas ^ this was the story you were working on i believe the other day.... can you flesh out how the *node specific* policy cases should work and try to include squeeds idea ?

Reply:
@squeed I incorporated your idea by adding a section about extending test scenarios by adding multiple nodes, putting pods to different subnets, and including host network pods.